File: frontend/ecotrack_app/lib/models/activity.dart
class Activity {
  final int? id;
  final String category;
  final String subtype;
  final double quantity;
  final String unit;
  final DateTime activityDate;
  final double? co2eKg;

  Activity({
    this.id,
    required this.category,
    required this.subtype,
    required this.quantity,
    required this.unit,
    required this.activityDate,
    this.co2eKg,
  });

  factory Activity.fromJson(Map<String, dynamic> json) {
    return Activity(
      id: json['id'],
      category: json['category'],
      subtype: json['subtype'],
      quantity: (json['quantity'] as num).toDouble(),
      unit: json['unit'],
      activityDate: DateTime.parse(json['activity_date']),
      co2eKg: json['co2e_kg'] != null ? (json['co2e_kg'] as num).toDouble() : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'category': category,
      'subtype': subtype,
      'quantity': quantity,
      'unit': unit,
      'activity_date': activityDate.toIso8601String(),
    };
  }
}

File: backend/app/models/activity.py
from datetime import datetime
from typing import Optional
from sqlmodel import SQLModel, Field
from pydantic import validator

class Activity(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="user.id")
    category: str # transport/energy/food/waste
    subtype: str
    quantity: float
    unit: str
    activity_date: datetime
    co2e_kg: float
    notes: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    @validator("quantity")
    def quantity_must_be_positive(cls, v):
        if v <= 0:
            raise ValueError("Quantity must be greater than zero.")
        return v
